File: logicx_biz/logicx_erp/report/virtual_bank_receipts_summary/virtual_bank_receipts_summary.py
import frappe
from frappe import _


def execute(filters=None):
	filters = frappe._dict(filters or {})
	return get_columns(), get_data(filters)


def get_columns():
	return [
		{
			"label": _("Mode of Payment"),
			"fieldname": "mode_of_payment",
			"fieldtype": "Link",
			"options": "Mode of Payment",
			"width": 200,
		},
		{
			"label": _("Total Paid Amount"),
			"fieldname": "total_paid_amount",
			"fieldtype": "Currency",
			"width": 160,
		},
		{
			# no data is selected for this column -- virtual_bank_receipts_summary.js's
			# formatter renders a button here for every row, using mode_of_payment off
			# the row and posting_date off this report's own filter (posting_date isn't
			# a per-row field here -- the whole report is already scoped to one date)
			"label": _("Breakup"),
			"fieldname": "open_breakup",
			"fieldtype": "HTML",
			"width": 110,
		},
	]


def get_data(filters):
	if not filters.get("posting_date"):
		frappe.throw(_("Posting Date is mandatory"))

	return frappe.db.sql(
		"""
		SELECT
			t.mode_of_payment,
			SUM(t.paid_amount) AS total_paid_amount
		FROM
		(
			SELECT
				pe.mode_of_payment,
				COALESCE(per.allocated_amount, pe.paid_amount) AS paid_amount
			FROM
				`tabPayment Entry` pe
			LEFT JOIN
				`tabPayment Entry Reference` per
					ON per.parent = pe.name
			WHERE
				pe.payment_type = 'Receive'
				AND pe.docstatus = 1
				AND pe.posting_date = %(posting_date)s

			UNION ALL

			SELECT
				sip.mode_of_payment,
				sip.amount AS paid_amount
			FROM
				`tabSales Invoice` si
			INNER JOIN
				`tabSales Invoice Payment` sip
					ON sip.parent = si.name
			WHERE
				si.docstatus = 1
				AND si.posting_date = %(posting_date)s
		) t
		GROUP BY
			t.mode_of_payment
		ORDER BY
			t.mode_of_payment
		""",
		{"posting_date": filters.get("posting_date")},
		as_dict=True,
	)

File: logicx_biz/logicx_erp/report/virtual_bank_receipts_summary/virtual_bank_receipts_summary.js
(function () {
	frappe.query_reports["Virtual Bank Receipts Summary"] = {
		onload: function (report) {
			add_breakup_buttons(report);
		},
		filters: [
			{
				fieldname: "posting_date",
				label: __("Posting Date"),
				fieldtype: "Date",
				default: frappe.datetime.add_days(frappe.datetime.get_today(), -1),
				reqd: 1,
			},
		],
		formatter: function (value, row, column, data, default_formatter) {
			if (column.fieldname === "open_breakup") {
				if (!data || !data.mode_of_payment) return "";
				const mode_of_payment = frappe.utils.escape_html(data.mode_of_payment);
				return (
					'<button type="button" class="btn btn-xs btn-default open-breakup-btn" ' +
					'data-mode-of-payment="' + mode_of_payment + '">' +
					__("Breakup") +
					"</button>"
				);
			}
			return default_formatter(value, row, column, data);
		},
	};

	function add_breakup_buttons(report) {
		// delegated so it survives the datatable re-rendering rows on every
		// refresh/filter change -- bind once against the page wrapper, which
		// persists for the life of the report
		report.page.wrapper.on("click", ".open-breakup-btn", function (e) {
			e.preventDefault();
			e.stopPropagation();
			const mode_of_payment = $(this).attr("data-mode-of-payment");
			if (!mode_of_payment) return;
			// posting_date isn't a per-row field here (it's this report's own
			// mandatory filter, applied identically to every row), so read it
			// live off the current report rather than off the clicked row
			const posting_date = frappe.query_report.get_filter_value("posting_date");
			frappe.route_options = {
				posting_date: posting_date,
				mode_of_payment: mode_of_payment,
			};
			frappe.set_route("query-report", "Virtual Bank Receipts Breakup");
		});
	}
})();
